### PR TITLE
sim-se: Implement free-list-based physical page allocator for SE mode

### LINKED ISSUE #1824: Large physical memory requirements because physical pages are not reallocated in SE mode
**Describe the bug**
Many programs require a much larger physical memory to run in gem5 than they require when run natively on the host. For example, I have noticed that 657.xz_s from the SPEC CPU2017 Integer benchmarks requires ~26GiB to run in gem5, but only ~15GiB to run on the host. This makes running many experiments in parallel difficult due to frequent out-of-memory errors on the host (because gem5 eats up so much RAM). Another consequence of having to run benchmarks with a larger physical memory in gem5 is that checkpoints occupy more disk space than they otherwise would.

I was able to root-cause the discrepancy between host vs. simulated physical memory requirements to gem5 not reallocating physical pages once they have been freed (e.g., via `munmap`). The physical page allocation code [here](https://github.com/gem5/gem5/blob/b1a44b89c7bae73fae2dc547bc1f871452075b85/src/sim/mem_pool.cc#L115) always allocates fresh pages rather than first looking to reallocate previously used but since freed ones.

Here's a proof-of-concept that only requires minimal physical memory on the host but unbounded memory in gem5 (`test.c`):
```c
#include <sys/mman.h>
#include <err.h>
#include <stdlib.h>

int main() {
  while (1) {
    char *map = mmap(NULL, 0x1000, PROT_READ, MAP_ANON | MAP_PRIVATE, -1, 0);
    if (map == MAP_FAILED)
      err(1, "mmap");
    volatile char c = *map;
    if (munmap(map, 0x1000) < 0)
      err(1, "munmap");
  }
}
```

**Affects version**
develop @ [ce4033e](https://github.com/gem5/gem5/commit/ce4033e36fa854b41fa566c27de3c24bd5e7ff5f)

**gem5 Modifications**
None

**To Reproduce**
1. Compile gem5: `scons build/X86/gem5.opt`
2. Build test program: `cc -o test test.c`
3. Run in KVM under SE: `./build/X86/gem5.opt configs/deprecated/example/se.py --cpu-type=X86KvmCPU --cmd ./test`

**Terminal Output**
```
gem5 Simulator System.  https://www.gem5.org
gem5 is copyrighted software; use the --copyright option for details.

gem5 version DEVELOP-FOR-24.1
gem5 compiled Nov 29 2024 09:44:11
gem5 started Nov 29 2024 15:52:14
gem5 executing on cafe-cet, pid 959023
command line: ./build/X86/gem5.opt configs/deprecated/example/se.py --cpu-type=X86KvmCPU --cmd ./test

warn: The se.py script is deprecated. It will be removed in future releases of  gem5.
Global frequency set at 1000000000000 ticks per second
warn: No dot file generated. Please install pydot to generate the dot file and pdf.
src/mem/dram_interface.cc:690: warn: DRAM device capacity (8192 Mbytes) does not match the address range assigned (512 Mbytes)
src/base/statistics.hh:279: warn: One of the stats is a legacy stat. Legacy stat is a stat that does not belong to any statistics::Group. Legacy stat is deprecated.
system.remote_gdb: Listening for connections on port 7000
src/mem/coherent_xbar.cc:140: warn: CoherentXBar system.membus has no snooping ports attached!
**** REAL SIMULATION ****
src/cpu/kvm/base.cc:169: info: KVM: Coalesced MMIO disabled by config.
src/cpu/kvm/base.cc:591: hack: Pretending totalOps is equivalent to totalInsts()
src/sim/simulate.cc:199: info: Entering event queue @ 0.  Starting simulation...
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x3a) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x48) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0xe1) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x12) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x11) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x4b564d01) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x4b564d00) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000000) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000001) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000020) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000021) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000022) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000023) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000100) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000101) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000102) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000103) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000104) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000105) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000003) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000002) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000010) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000080) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x400000b0) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000073) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000106) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000107) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000108) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x40000118) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x400000ff) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x400000f1) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x400000f2) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x400000f3) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x400000f4) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x400000f5) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x4b564d02) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x4b564d03) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x4b564d04) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x4b564d06) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x4b564d07) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x3b) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x6e0) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x10a) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x345) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x1a0) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x4d0) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x9e) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x34) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0xce) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x140) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x1fc) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x8b) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x480) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x48d) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x48e) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x48f) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x490) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x485) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x486) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x488) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x48a) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x48b) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x48c) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x491) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0xc0010015) unsupported by gem5. Skipping.
src/arch/x86/kvm/x86_cpu.cc:1688: warn: kvm-x86: MSR (0x4b564d05) unsupported by gem5. Skipping.
src/sim/mem_state.cc:448: info: Increasing stack size by one page.
src/sim/syscall_emul.cc:86: warn: ignoring syscall mprotect(...)
src/sim/syscall_emul.cc:86: warn: ignoring syscall set_robust_list(...)
src/sim/syscall_emul.cc:86: warn: ignoring syscall rseq(...)
src/sim/syscall_emul.cc:86: warn: ignoring syscall mprotect(...)
src/sim/syscall_emul.cc:86: warn: ignoring syscall mprotect(...)
src/sim/syscall_emul.cc:86: warn: ignoring syscall mprotect(...)
src/sim/mem_pool.cc:120: fatal: fatal condition freePages() <= 0 occurred: Out of memory, please increase size of physical memory.
Memory Usage: 650384 KBytes
```

**Expected behavior**
Ideally, benchmarks like 657.xz_s take around the same amount of physical memory to simulate as they do to run on the host.

As for the POC/test, it should run indefinitely and never run out of memory.

**Host Operating System**
Ubuntu 22.04

**Host ISA**
X86

**Compiler used**
GCC 12

**Additional information**
I'll submit a patch for reallocating freed physical pages.

### PR DESCRIPTION
**Pull request by nmosier**
This series of commits implements a free-list-based physical page allocator for SE mode, supported by the following:
* 700a5a5 Adds a new stat, `maxAllocatedBytes`, to memory pools, which tracks the maximum physical memory, in bytes, that has been allocated at any point in the simulation so far.
* 6cfa219 Fixes an issue with `statistics::Group` that made gem5 crash after making `MemPool` a stats group
* 385fc6e Implements the free-list physical page allocator itself.

These patches fix issue #1809.

I evaluated the old, naive physical page allocator ("old" in the table) versus the new, free-list-based physical page allocator ("new" in the table) on the SPEC CPU2017 Integer benchmarks, using the X86KvmCPU model in SE mode. I collected three stats:
* **sim mem**: the physical memory allocated in the simulator (given in the `maxAllocatedBytes` stats), labeled as "sim mem" in the table
* **host mem**: the maximum memory usage of the gem5 process (collected via `/usr/bin/time -v`)
* **host time**: the host runtime to run the benchmark to completion.
<table>
    <tr>
        <th rowspan="2">benchmark</th>
        <th colspan="2">sim mem (maxAllocatedBytes, in GiB)</th>
        <th colspan="2">host mem (GiB)</th>
        <th colspan="2">host runtime (s)</th>
    </tr>
    <tr>
        <th>old</th><th>new</th>
        <th>old</th><th>new</th>
        <th>old</th><th>new</th>
    </tr>
    <tr><td>perlbench.0</td><td>0.20</td><td>0.20</td><td>0.28</td><td>0.29</td><td>74.91</td><td>75.43</td></tr>
    <tr><td>perlbench.1</td><td>0.16</td><td>0.15</td><td>0.24</td><td>0.23</td><td>43.86</td><td>43.88</td></tr>
    <tr><td>perlbench.2</td><td>0.09</td><td>0.09</td><td>0.17</td><td>0.17</td><td>69.68</td><td>71.45</td></tr>
    <tr><td>gcc.0</td><td>9.29</td><td>7.48</td><td>9.47</td><td>7.67</td><td>344.63</td><td>391.42</td></tr>
    <tr><td>gcc.1</td><td>3.24</td><td>3.05</td><td>3.36</td><td>3.18</td><td>101.30</td><td>100.75</td></tr>
    <tr><td>gcc.2</td><td>3.29</td><td>3.04</td><td>3.42</td><td>3.17</td><td>99.43</td><td>100.97</td></tr>
    <tr><td>mcf</td><td>13.22</td><td>4.83</td><td>12.40</td><td>4.96</td><td>429.66</td><td>477.23</td></tr>
    <tr><td>omnetpp</td><td>0.27</td><td>0.24</td><td>0.36</td><td>0.33</td><td>214.43</td><td>214.22</td></tr>
    <tr><td>xalancbmk</td><td>0.47</td><td>0.47</td><td>0.58</td><td>0.58</td><td>129.34</td><td>129.84</td></tr>
    <tr><td>x264.0</td><td>0.15</td><td>0.15</td><td>0.23</td><td>0.23</td><td>16.95</td><td>17.10</td></tr>
    <tr><td>x264.1</td><td>0.04</td><td>0.04</td><td>0.12</td><td>0.12</td><td>44.75</td><td>44.60</td></tr>
    <tr><td>x264.2</td><td>0.15</td><td>0.15</td><td>0.23</td><td>0.23</td><td>46.70</td><td>46.64</td></tr>
    <tr><td>deepsjeng</td><td>6.73</td><td>6.73</td><td>6.89</td><td>6.89</td><td>250.32</td><td>250.02</td></tr>
    <tr><td>leela</td><td>0.10</td><td>0.03</td><td>0.18</td><td>0.10</td><td>237.70</td><td>237.85</td></tr>
    <tr><td>exchange2</td><td>0.00</td><td>0.00</td><td>0.08</td><td>0.08</td><td>319.68</td><td>320.10</td></tr>
    <tr><td>xz.0</td><td>26.06</td><td>15.03</td><td>26.54</td><td>15.51</td><td>592.75</td><td>656.06</td></tr>
    <tr><td>xz.1</td><td>6.15</td><td>4.11</td><td>6.51</td><td>4.48</td><td>1173.43</td><td>1182.33</td></tr>
    <tr><td>total</td><td>69.61</td><td>45.80</td><td>71.06</td><td>48.22</td><td>4189.52</td><td>4359.89</td></tr>
    <tr><td>total, normalized</td><td>1</td><td>0.66</td><td>1</td><td>0.68</td><td>1</td><td>1.04</td></tr>
</table>

The results show that using the new free-list-based physical page allocator significantly reduces the amount of RAM required to run the SPEC benchmarks in gem5 in SE mode. There is a 4% increase in runtime overhead. This is simply due to the need to zero out physical pages upon deallocation. Notably, the runtime is only significantly slower for benchmarks that free a lot of memory (e.g., xz.0 and xz.1) and thus have significantly reduced memory requirements — that is, we pay for reducing memory overhead with a slight increase in runtime overhead. However, this runtime overhead becomes negligible for simulated CPUs (rather than KVM-based CPUs).

### CODE REVIEW

**Comment by giactra** [comment]
Hi @nmosier; this looks like a pretty useful change.

I have two initial thoughts

1) Can't we make page zeroing an optional behaviour? In most cases we are more concerned about simulation speed rather than modelling the kernel zeroing behaviour (which happens at allocation anyway and not deallocation)

2) Could you elaborate a bit further on the stats change? You say the problem happens when we destruct a stats group. But this shouldn't really happen as we might lose important information. Is it a problem now because we deallocate MemPool? Wouldn't it make more sense to add the stats to MemPools class instead?

**Comment by erin-le** [comment]
@nmosier Hello, thank you for your contribution! When you have time, could you respond to the comments above?

**Comment by halfgodxzc** [comment]
> @nmosier Hello, thank you for your contribution! When you have time, could you respond to the comments above?

@erin-le  Hi, I’ve reviewed the commit thoroughly and am very interested in the changes. This is my first time participating in an issue, and I wanted to ask if it would be acceptable for me to contribute by writing code to address the concerns mentioned in the comments.

**Comment by erin-le** [comment]
@halfgodxzc It doesn't look like the author is actively working on this PR, so please go ahead!

**Comment by nmosier** [comment]
@giactra Thanks for taking a look and the feedback. I addressed your comments and submitted a revised patch.
1. I made page zeroing optional by adding the `zeroPages` param to the `Process` python object. Generally, zeroing pages is necessary because glibc malloc assumes that all newly mmap'ed pages are zero-filled. See issue #342 and pull request #343 for more details.
2. I chose to zero pages out upon deallocation rather than allocation to avoid unnecessarily zeroing out pages that have never been allocated before and thus are already zero-filled. I added a comment at src/sim/process.cc:362 explaining this.
3. I moved the `maxAllocatedBytes` stats from `MemPool` to `MemPools` as you suggested, and was thus able to avoid mucking around with stats group destructors. The new patch does not modify anything in src/base now.

File: src/base/free_list.cc
Comment (powerjg) [inline src/base/free_list.cc:1]: Why is this empty file needed? Shouldn't the code in `free_list.hh` be compiled when it is included in `process.cc`? Also, it's not compiled in the `SConscript`.

File: src/sim/Process.py [hunk 1]
```diff
@@ -51,6 +51,10 @@ def map(self, vaddr, paddr, size, cacheable=False):
     )
     kvmInSE = Param.Bool("false", "initialize the process for KvmCPU in SE")
     maxStackSize = Param.MemorySize("64MiB", "maximum size of the stack")
+    zeroPages = Param.Bool(
+        True,
+        "ensure all allocated pages are zero-filled. glibc malloc generally requires this. Disable at your own risk.",
```
Comment (powerjg) [inline src/sim/Process.py]: This line is beyond 80 characters

File: src/sim/mem_pool.cc [hunk 2]
```diff
@@ -114,31 +95,70 @@ MemPool::totalBytes() const
...
+
+    Counter free_page_num;
+    if (optParamIn(cp, "free_page_num", free_page_num)) {
+        // Handle checkpoints generated by legacy page allocator
+        // for backwards compatibility.
+        warn("Checkpoint contains legacy MemPool state (free_page_num)\n");
+        freePhysPages.insert(free_page_num, _totalPages - free_page_num);
+    } else {
+        ScopedCheckpointSection sec(cp, "free_list");
+        int n;
```
Comment (powerjg) [inline src/sim/mem_pool.cc]:
```suggestion
        int n = 0;
```

Just to make sure the static analysis tools won't report this as an uninitialized variable.

**Review by powerjg — COMMENTED** [review]
Thanks for the updates. Overall, this looks good to me. I have a couple of small things below.

File: src/base/free_list.cc
Comment (nmosier) [reply src/base/free_list.cc:1]: Ah, not sure why I added that. I removed it in the latest patch.

File: src/sim/Process.py (hunk 1 above)
Comment (nmosier) [reply src/sim/Process.py]: Oops, fixed!

File: src/sim/mem_pool.cc (hunk 2 above)
Comment (nmosier) [reply src/sim/mem_pool.cc]: Fixed!

**Comment by nmosier** [comment]
Thanks! I just addressed your three comments with my latest patch.

**Comment by powerjg** [comment]
@giactra can you take a look at this after the response from @nmosier

File: src/base/free_list.test.cc [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <gtest/gtest.h>
+
+#include "base/free_list.hh"
+
+using namespace gem5;
+
+TEST(FreeListTest, Simple)
```
Comment (giactra) [inline src/base/free_list.test.cc:36]: Adding these unit tests is great but I feel they are not stressing some tricky corner cases, like for example what happens if you start inserting overlapping ranges

Judging from https://github.com/gem5/gem5/pull/1825/files#diff-544445642128ae52856649383a6912f556c5de8ea74af7400893c82c4b6b6b3bR74 I have the feeling this is not covered by your FreeList implementation

File: src/base/free_list.hh [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        // Insert new range.
+        _ranges.emplace(it, base, size);
+    }
+
+    /**
+     * Allocate a region of size @param size out of free space.
+     * @return whether the allocation succeeded.
+     */
+    bool
+    allocate(T size, T& base)
```
Comment (giactra) [inline src/base/free_list.hh]: Nit, how do you feel about:

```
std::optional<T>
allocate(T size)
{
```

In this way you don't have to pass the return value by reference

File: src/sim/mem_pool.cc [hunk 5]
```diff
@@ -148,10 +168,30 @@ MemPools::populate(const AddrRangeList &memories)
...
+
 Addr
 MemPools::allocPhysPages(int npages, int pool_id)
 {
-    return pools[pool_id].allocate(npages);
+    const Addr addr = pools[pool_id].allocate(npages);
+
+    // Recompute max allocated bytes stat, since it may have changed.
+    stats.maxAllocatedBytes =
+        std::max<Addr>(stats.maxAllocatedBytes.value(), allocatedBytes());
```
Comment (giactra) [inline src/sim/mem_pool.cc]: I think this needs a bit of attention

When dealing with stats one always needs to take into consideration they can (and will) be reset manually with the m5 reset pseudo op before entering a ROI.

So at a certain point maxAllocatedBytes might be reset to 0.

What is the stat supposed to model? Is it supposed to model the maximum number of allocated bytes in the entire simulation (m5 reset shouldn't affect it)? Or do we want to model the number of allocated bytes in a specific time window (m5 reset should affect it)

**Review by giactra — COMMENTED** [review]
Thanks @nmosier for your changes! Please check my comments below

File: src/sim/mem_pool.cc (hunk 5 above)
Comment (powerjg) [reply src/sim/mem_pool.cc]: > m5 reset shouldn't affect it

Let's take this discussion off of this thread, but I'm not sure we should have any "stats" which are not reset. We've had confusion in the past with this. Maybe we should have a different word/interface for stats which don't reset.

See #1615 and #1699 for examples.

Let's discuss this at a high level at the next developers meeting.

**Comment by erin-le** [comment]
Hello @nmosier, when you have time, could you address these comments? Thank you!

**Comment by erin-le** [comment]
@nmosier Hello, just another reminder about the comments on this PR. Thank you for your contribution!

File: src/base/free_list.test.cc (hunk 3 above)
Comment (nmosier) [reply src/base/free_list.test.cc:36]: Correct, the free list doesn't support inserting overlapping ranges, i.e., it doesn't permit double frees. I added a few extra test cases to verify that the free list panics when a double free is attempted.

File: src/base/free_list.hh (hunk 4 above)
Comment (nmosier) [reply src/base/free_list.hh]: Sure, makes sense. I made that change.

File: src/sim/mem_pool.cc (hunk 5 above)
Comment (nmosier) [reply src/sim/mem_pool.cc]: I removed this stat to simplify the patch, since it's not necessary for the free list to work properly.

**Comment by nmosier** [comment]
@giactra I pushed a new version that tries to address your feedback. Apologies for the delay.

**Review by giactra — APPROVED** [review]
Thanks @nmosier for your changes!
